src/main/kotlin/org/rust/lang/refactoring/RsImportOptimizer.kt

```diff
@@ -72,18 +71,63 @@ class RsImportOptimizer : ImportOptimizer {
             val first = mod.childrenOfType<RsElement>()
                 .firstOrNull { it.textOffset >= offset && it !is RsExternCrateItem && it !is RsAttr } ?: return
             val psiFactory = RsPsiFactory(mod.project)
-            val sortedUses = uses
-                .sortedBy { it.useSpeck?.pathText }
-                .mapNotNull { it.copy() as? RsUseItem }
-            sortedUses
-                .mapNotNull { it.useSpeck }
-                .forEach { optimizeUseSpeck(psiFactory, it) }
-            for (importPath in sortedUses) {
-                mod.addBefore(importPath, first)
+            val sortedUsesGroups = uses
+                .map { UseItemWrapper(it) }
+                .groupBy({ it.packageGroupLevel }, { it })
+                .map { (groupLevel, useItemWrapper) ->
+                    groupLevel to useItemWrapper
+                        .sortedBy { it.useSpeckText }
+                        .mapNotNull { it.useItem.copy() as? RsUseItem }
+                }
+                .filter { it.second.isNotEmpty() }
+                .sortedBy { it.first }
+            sortedUsesGroups.forEach { sortedUses ->
+                sortedUses.second
+                    .mapNotNull { it.useSpeck }
+                    .forEach { optimizeUseSpeck(psiFactory, it) }
+            }
+
+            for ((_, sortedUses) in sortedUsesGroups) {
+                val lastUseItem = sortedUses.map { importPath ->
```

Comment:
You don't need `map` here, it can be replaced by `for` loop and mutable variable.
It allows avoiding unnecessary allocation

Reply:
Done